src/appdistribution/yaml_helper.ts

```diff
@@ -0,0 +1,103 @@
+import * as jsYaml from "js-yaml";
+import { getErrMsg, FirebaseError } from "../error";
+import { TestCase } from "./types";
+
+declare interface YamlStep {
+  goal?: string;
+  hint?: string;
+  successCriteria?: string;
+}
+
+const ALLOWED_YAML_STEP_KEYS = new Set(["goal", "hint", "successCriteria"]);
+
+declare interface YamlTestCase {
+  displayName?: string;
+  id?: string;
+  prerequisiteTestCaseId?: string;
+  steps?: YamlStep[];
+}
+
+const ALLOWED_YAML_TEST_CASE_KEYS = new Set([
+  "displayName",
+  "id",
+  "prerequisiteTestCaseId",
+  "steps",
+]);
+
+function extractIdFromResourceName(name: string): string {
+  return name.split("/").pop() ?? "";
+}
+
+function toYamlTestCases(testCases: TestCase[]): YamlTestCase[] {
+  return testCases.map((testCase) => ({
+    displayName: testCase.displayName ?? "",
+    id: extractIdFromResourceName(testCase.name!),
```

Comment:
Prefer avoiding nonnull assertions - usually they mean that our typing is inaccurate. A null fallback would be preferable.

Reply:
Unfortunately I can't fix this with the typing:

For a new `TestCase` that is being created, the resource name can be undefined, because according to the [relevant API](http://go/aip/133#user-specified-ids) "the name field on the resource **must** be ignored." .

But any `TestCase` returned by the server (in response to a create, get, list, or update), this will be set.

All the test cases passed to this function have been retrieved from the server, so the `name` will be set. I've added a comment explaining this.